src/deriver/ScopedKeys.js

```diff
@@ -0,0 +1,105 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const HKDF = require('node-hkdf');
+const base64url = require('base64url');
+
+const KEY_LENGTH = 32;
+
+/**
+ * Scoped key deriver
+ * @desc Used by the Firefox Accounts content server
+ * @module deriver-ScopedKeys
+ * @example
+ * ```js
+ * const scopedKeys = new fxaCryptoDeriver.ScopedKeys();
+ *
+ * return scopedKeys.deriveScopedKeys({
+ *   identifier: 'https://identity.mozilla.com/apps/notes',
+ *   inputKey: 'bc3851e9e610f631df94d7883d5defd5e5f55ab520bd5a9ae33dae26575c6b1a',
+ *   keyMaterial: '0000000000000000000000000000000000000000000000000000000000000000',
+ *   timestamp: 1494446722583
+ * });
+ * ```
+ */
+class ScopedKeys {
+  /**
+   * Derive a scoped key
+   * @method deriveScopedKeys
+   * @param {object} options - required set of options to derive a scoped key
+   * @param {string} options.inputKey - input key that the scoped key is derived from
+   * @param {string} options.keyMaterial - a 32-byte string of additional entropy specific
+   * @param {number} options.timestamp - a 10-digit number, the timestamp at which this scoped key most recently changed
+   * @param {string} options.identifier - a unique URI identifying the requested scoped key
+   * @returns {Promise}
+   */
+  deriveScopedKeys(options) {
+    return new Promise((resolve) => {
+      if (! options.inputKey) {
+        throw new Error('inputKey required');
+      }
+
+      if (! options.keyMaterial) {
+        throw new Error('keyMaterial required');
+      }
+
+      if (! options.timestamp) {
+        throw new Error('timestamp required');
+      }
+
+      if (! options.identifier) {
+        throw new Error('identifier required');
+      }
+
+      const context = 'identity.mozilla.com/picl/v1/scoped_key\n' +
+        options.identifier;
+      const contextKid = 'identity.mozilla.com/picl/v1/scoped_kid\n' +
+        options.identifier;
+      const scopedKey = {
+        kty: 'oct',
+        scope: options.identifier,
+      };
+
+      this._deriveHKDF(options.keyMaterial, options.inputKey, context)
+        .then((key) => {
+          scopedKey.k = base64url(key);
+
+          return this._deriveHKDF(options.keyMaterial, options.inputKey, contextKid);
+        })
+        .then((kidKey) => {
+          const keyTimestamp = Math.round(options.timestamp / 1000);
+
+          scopedKey.kid = keyTimestamp + '-' + base64url(kidKey);
```

Comment:
I wonder if we should do some sanity-checking here that the stringified timestamp as the right number of digits, and either zero-pad if it's unexpectedly short, or error out if it's unexpectedly long.

Reply:
Specifically, my concern is with maintaining proper sort order of the stringified key id, which depends on the timestamp field taking up a known number of characters.

Reply:
👍

Reply:
Done.